### PR TITLE
Implement std::iter::once/empty (RFC 771)

### LINKED ISSUE #24443: Tracking issue for Std::iter::{once, empty} (RFC 771)
https://github.com/rust-lang/rfcs/pull/771

### PR DESCRIPTION
**Pull request by XMPPwocky**
Closes #24443.

### CODE REVIEW

File: src/libcore/iter.rs [hunk 1]
```diff
@@ -3030,6 +3030,45 @@ pub fn repeat<T: Clone>(elt: T) -> Repeat<T> {
...
+#[unstable(feature="core", reason = "new addition")]
+pub fn empty<T>() -> Empty<T> {
+    Empty(marker::PhantomData)
+}
+
+
+/// An iterator that yields an element exactly once.
+#[unstable(feature="core", reason = "new addition")]
+pub struct Once<T> {
+    inner: ::option::IntoIter<T>
```
Comment (tbu-) [inline src/libcore/iter.rs]: I guess a `use option` at the beginning would be nicer than the global path here.

**Comment by bluss** [comment]
These iterators should implement `Clone`, `DoubleEndedIterator` and `ExactSizeIterator` too.

File: src/libcore/iter.rs [hunk 2]
```diff
@@ -3030,6 +3030,45 @@ pub fn repeat<T: Clone>(elt: T) -> Repeat<T> {
...
+/// An iterator that yields nothing.
+#[unstable(feature="core", reason = "new addition")]
+pub struct Empty<T>(marker::PhantomData<T>);
+
+impl<T> Iterator for Empty<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<T> {
+        None
+    }
```
Comment (bluss) [inline src/libcore/iter.rs]: size_hint is important even when it's trivial.

File: src/libcore/iter.rs (hunk 1 above)
Comment (Stebalien) [inline src/libcore/iter.rs]: `::option::IntoIter<T>` isn't really necessary. You could just use `Option` + `take()`.

**Comment by Stebalien** [comment]
IMHO, they should also implement Default so that structs that include them can derive Default.

File: src/libcore/iter.rs [hunk 3]
```diff
@@ -3030,6 +3030,45 @@ pub fn repeat<T: Clone>(elt: T) -> Repeat<T> {
     Repeat{element: elt}
 }
 
+/// An iterator that yields nothing.
+#[unstable(feature="core", reason = "new addition")]
```
Comment (alexcrichton) [inline src/libcore/iter.rs]: Could this use a different feature name than "core" here as well? Something like `iter_empty` and `iter_once` should be fine. Also, can you tag the `impl` blocks below with `#[unstable]` as well?

File: src/libcore/iter.rs (hunk 1 above)
Comment (alexcrichton) [inline src/libcore/iter.rs]: I would recommend using `IntoIter` (or some other iterator adapator) as it allows proxying the `next`, `next_back`, and `size_hint` functions (primarily `size_hint`)

**Comment by alexcrichton** [comment]
I agree with @bluss that it'd be nice to see the other common iterator traits implemented for these primitives as well (they should be pretty easy).

File: src/libcore/iter.rs (hunk 1 above)
Comment (Stebalien) [inline src/libcore/iter.rs]: That's three levels of indirection (IntoIter -> Item -> Option) but I guess they'll just get optimized away anyways.

**Comment by XMPPwocky** [comment]
@alexcrichton This good?

**Comment by Stebalien** [comment]
You could also impl Default on `once<T>()` but I'm ambivalent.

File: src/libcore/iter.rs [hunk 4]
```diff
@@ -3030,6 +3030,100 @@ pub fn repeat<T: Clone>(elt: T) -> Repeat<T> {
...
+/// Creates an iterator that yields nothing.
+#[unstable(feature="iter_empty", reason = "new addition")]
+pub fn empty<T>() -> Empty<T> {
+    Empty(marker::PhantomData)
+}
+
+/// An iterator that yields an element exactly once.
+#[unstable(feature="iter_once", reason = "new addition")]
+pub struct Once<T> {
+    inner: ::option::IntoIter<T>
```
Comment (tbu-) [inline src/libcore/iter.rs]: >I guess a `use option` at the beginning would be nicer than the global path here.

**Comment by bluss** [comment]
Thank you @XMPPwocky I've needed these iterators!

File: src/libcore/iter.rs (hunk 4 above)
Comment (XMPPwocky) [inline src/libcore/iter.rs]: I'd do that if I used it twice, but adding an import to a 3000+ line file to save two colons on one line seemed excessive.

Comment (tbu-) [inline src/libcore/iter.rs]: It's unusual to use a global path, though.

**Comment by alexcrichton** [comment]
This looks good to me, thanks @XMPPwocky! r=me with a squash

File: src/libcore/lib.rs [hunk 5]
```diff
@@ -167,6 +167,7 @@ mod core {
     pub use marker;
     pub use option;
     pub use iter;
+    pub use default;
```
Comment (alexcrichton) [inline src/libcore/lib.rs]: Is this still necessary?

Comment (XMPPwocky) [inline src/libcore/lib.rs]: Looks like it's not.

**Comment by alexcrichton** [comment]
@bor: r+ 103e79d

File: src/libcore/iter.rs (hunk 4 above)
Comment (cristicbz) [inline src/libcore/iter.rs]: Alternatively just store Option<T> and return self.inner.take() on next?

File: src/libcoretest/iter.rs [hunk 6]
```diff
@@ -1096,6 +1096,19 @@ fn test_fuse_count() {
...
+fn test_once() {
+    let mut it = once(42);
+    assert_eq!(it.next(), Some(42));
+    assert_eq!(it.next(), None);
+}
+
+#[test]
+fn test_empty() {
+    let mut it = empty::<i32>();
+    assert_eq!(it.next(), None);
```
Comment (cristicbz) [inline src/libcoretest/iter.rs]: Maybe add another assert_eq to check that the iterators are fused? And above for once.

**Pull request merged**